rclc/src/functions.c

```diff
@@ -0,0 +1,347 @@
+
+#define PRINT_RCL_ERROR(rclc, rcl) \
+  do { \
+    fprintf(stderr, "[" #rclc "] error in " #rcl ": %s\n", rcl_get_error_string_safe()); \
+    rcl_reset_error(); \
+  } while (0)
+
+rclc_ret_t
+rclc_init(int argc, char ** argv)
+{
+  rcl_ret_t rc = rcl_init(argc, argv, rcl_get_default_allocator());
+  if (rc != RCL_RET_OK) {
+    PRINT_RCL_ERROR(rclc_init, rcl_init);
+  }
+  return rc;
+}
+
+bool
+rclc_ok(void)
+{
+  return rcl_ok();
+}
+
+static
+inline
+void
+_rclc_spin_node_exit(rcl_wait_set_t * wait_set)
+{
+  rcl_ret_t rc = rcl_wait_set_fini(wait_set);
+  if (rc != RCL_RET_OK) {
+    PRINT_RCL_ERROR(rclc_spin_node, rcl_wait_set_fini);
+  }
+}
+
+void
+rclc_spin_node_once(rclc_node_t * node, size_t timeout_ms)
+{
+  // [FIXME] Make the wait_set not empty if there is no subscription
+  const size_t dummy = node->subs_s ? 0 : 1;
+
+  rcl_wait_set_t wait_set = rcl_get_zero_initialized_wait_set();
+  rcl_ret_t rc =
+    rcl_wait_set_init(&wait_set, node->subs_s, 0, dummy, 0, 0, rcl_get_default_allocator());
+  if (rc != RCL_RET_OK) {
+    PRINT_RCL_ERROR(rclc_spin_node, rcl_wait_set_init);
+    return;
+  }
+
+  rc = rcl_wait_set_clear_subscriptions(&wait_set);
+  if (rc != RCL_RET_OK) {
+    PRINT_RCL_ERROR(rclc_spin_node, rcl_wait_set_clear_subscriptions);
+    _rclc_spin_node_exit(&wait_set);
+    return;
+  }
+
+  for (size_t i = 0; i < node->subs_s; ++i) {
+    rc = rcl_wait_set_add_subscription(&wait_set, &node->subs[i]->rcl_subscription);
+    if (rc != RCL_RET_OK) {
+      PRINT_RCL_ERROR(rclc_spin_node, rcl_wait_set_add_subscription);
+      _rclc_spin_node_exit(&wait_set);
+      return;
+    }
+  }
+
+  rc = rcl_wait(&wait_set, RCL_MS_TO_NS(timeout_ms));
+  if (rc == RCL_RET_TIMEOUT) {
+    _rclc_spin_node_exit(&wait_set);
+    return;
+  }
+
+  if (rc != RCL_RET_OK) {
+    PRINT_RCL_ERROR(rclc_spin_node, rcl_wait);
+    _rclc_spin_node_exit(&wait_set);
+    return;
+  }
+
+  for (size_t i = 0; i < wait_set.size_of_subscriptions; ++i) {
+    rclc_subscription_t * sub = NULL;
+
+    for (size_t j = 0; j < node->subs_s; ++j) {
+      if (&node->subs[j]->rcl_subscription == wait_set.subscriptions[i]) {
+        sub = node->subs[j];
+      }
+    }
+
+    if (sub == NULL) {
+      fprintf(stderr, "[rclc_spin_node] unable to find subscription in node.\n");
+      _rclc_spin_node_exit(&wait_set);
+    }
+
+    void * msg = ZERO_ALLOCATE(sub->type_support.size_of);
+
+    rc = rcl_take(wait_set.subscriptions[i], msg, NULL);
+    if (rc != RCL_RET_OK) {
+      PRINT_RCL_ERROR(rclc_spin_node, rcl_take);
+      _rclc_spin_node_exit(&wait_set);
+      return;
+    }
+
+    sub->user_callback(msg);
+  }
+
+  _rclc_spin_node_exit(&wait_set);
+}
+
+void
+rclc_spin_node(rclc_node_t * node)
+{
+  rcl_wait_set_t wait_set = rcl_get_zero_initialized_wait_set();
+  rcl_ret_t rc =
+    rcl_wait_set_init(&wait_set, node->subs_s, 0, 0, 0, 0, rcl_get_default_allocator());
```

Comment:
Maybe we should add the [dummy hack](https://github.com/ros2/rclc/pull/8/files#diff-2ac3bdc884a04dc4cdc95db51221e264R70) here as well as the passed node may not have any subscription